src/compiler/checker.ts

```diff
@@ -3993,6 +4027,37 @@ namespace ts {
             return links.resolvedType;
         }
 
+        // We do not perform supertype reduction on intersection types. Intersection types are created only by the &
+        // type operator and we can't reduce those because we want to support recursive intersection types. For example,
+        // a type alias of the form "type List<T> = T & { next: List<T> }" cannot be reduced during its declaration.

```

Comment:
Actually I misunderstood. Recursive union types do work, as long as the type alias reference is not in an area that is reachable when we process the type (for example in an object type). But that only works because we pass true for noSubtypeReduction in getTypeFromUnionTypeNode. In that case, the fact that we want recursive intersection types doesn't preclude us from supporting supertype reduction. It just means that getTypeFromIntersectionTypeNode cannot trigger supertype reduction. But presumably, getApparentType should trigger supertype reduction, no?